design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,503 @@
+To accommodate the ippools, capture and nmstate new fields will be needed there
+the following is an example of the install-config.yaml from openshift with
+the new fields:
+
+```yaml
+apiVersion: v1
+baseDomain: test.metalkube.org
+metadata:
+  name: ostest
+platform:
+  baremetal:
+    networkPolicy:
+      ipPools:
+        traffic-1:
+          cidr: 10.10.10.0/24
+          start: 10.10.10.1
+          end: 10.10.10.10
+          exclude: ["10.10.10.4", "10.10.10.7"]
+        traffic-2:
+          cidr: 10.10.11.0/24
+          start: 10.10.11.1
+          end: 10.10.10.11.10
+          exclude: ["10.10.11.4", "10.10.11.7"]
+      capture:
+        address-from-traffic-1: ippool.traffic-1.allocate
+        address-from-traffic-2: ippool.traffic-2.allocate
+      desiredState:
+        interfaces:
+        - name: eth1
+          type: ethernet
+          state: up
+          ipv4:
+            address:
+            - ip: "{{ capture.address-from-foo.ip }}"
+              prefix-length: "{{ capture.address-from-foo.prefix-length }}"
+            dhcp: false
+            enabled: true
+        - name: eth2
+          type: ethernet
+          state: up
+          ipv4:
+            address:
+            - ip: "{{ capture.address-from-bar.ip }}"
+              prefix-length: "{{ capture.address-from-bar.prefix-length }}"
+            dhcp: false
+            enabled: true
+    hosts:
+      - name: openshift-master-0
+        networkData: [static network configuration for master-0]
+      - name: openshift-worker-0
+        networkData: [static network configuration for worker-0]
+```
+
+The new field `platform.baremetal.networkPolicy` will contain the
+content of the day 2 `NodeNetworkConfigurationPolicy` and also the `ipPools`
+since there is not going to be multiple `networkPolicy` fields.
+
+### Implementation Details/Notes/Constraints
+
+#### Day 2
+
+Since kubernetes-nmstate has a distributed architecture each of the handler
+pods will calculate an allocatable ip from the `NodeIPPool` status and
+retry the process on conflict
+
+Handlers will check NNCEs at Reconcile and if the ip has being already assigned
+it will re-use that and update `NodeIPPool` status with it, this way the
+`NodeIPPool` can be reset but IPs are kept.
+
+To optimize on storage and lookup instead of using a list of allocated offset
+it's possible to use a bitmap, the OVN project use that approach to store the
+offsets on a range [1], a snapshot of the bitmap can be stored as base64 at the
+`NodeIPPool` status.
+
+For the scale down scenario, the current kubernetes-nmstate architecture
+contains a controller for node that will Reconcile on node deletion, but since
+the nodes may be drained before deletion the pod running the controller
+will no longer be running, this means that the controller may need to run at
+control-plane nodes and watch for all the nodes.
+
+[1] https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ipallocator/allocator.go
```

Comment:
Does this section make sense here ? Like it seems to not fit - at all - between the `Baremetal IPI` and the `API Extensions baremetal IPI` ? It seems to be a simple general purpose implementation detail. 

It totally broke my concentration on anything related to baremetal.

Reply:
Going to rework the doc to remove baremetal IPI, that has to be done at a follow up.